### PR TITLE
DCS Continued

### PR DESCRIPTION
**Pull request by Cyberboss**
I hate git sometimes

Closes #29317 

This PR adds documentation and final touches missing from #29178 

# Datum Component System (DCS)

## Concept

Loosely adapted from /vg/. This is an entity component system for adding behaviours to datums when inheritance doesn't quite cut it. By using signals and events instead of direct inheritance, you can inject behaviours without hacky overloads. It requires a different method of thinking, but is not hard to use correctly. If a behaviour can have application across more than one thing. Make it generic, make it a component. Atom/mob/obj event? Give it a signal, and forward it's arguments with a `SendSignal()` call. Now every component that want's to can also know about this happening.

#### Slippery things

At the time of this writing, every object that is slippery overrides atom/Crossed does some checks, then slips the mob. Instead of all those Crossed overrides they could add a slippery component to all these objects. And have the checks in one proc that is run by the Crossed event

#### Powercells

A lot of objects have powercells. The `get_cell()` proc was added to give generic access to the cell var if it had one. This is just a specific use case of `GetComponent()`

#### Radios

The radio object as it is should not exist, given that more things use the _concept_ of radios rather than the object itself. The actual function of the radio can exist in a component which all the things that use it (Request consoles, actual radios, the SM shard) can add to themselves.

#### Standos

Stands have a lot of procs which mimic mob procs. Rather than inserting hooks for all these procs in overrides, the same can be accomplished with signals

### Vars

1. `/datum/var/list/datum_components` (private)
    * Lazy list of all components a datum has   (TODO: Make this a typecache with longer paths overwriting shorter ones maybe? It'd be weird)
1. `/datum/component/var/enabled` (protected, boolean)
    * If the component is enabled. If not, it will not react to signals
    * TRUE by default
1. `/datum/component/var/dupe_mode` (protected, enum)
    * How multiple components of the exact same type are handled when added to the datum.
        * `COMPONENT_DUPE_HIGHLANDER` (default): Old component will be deleted, new component will first have `/datum/component/proc/InheritComponent(datum/component/old, FALSE)` on it
        * `COMPONENT_DUPE_ALLOWED`: The components will be treated as seperate, `GetComponent()` will return the first added
        * `COMPONENT_DUPE_UNIQUE`: New component will be deleted, old component will first have `/datum/component/proc/InheritComponent(datum/component/new, TRUE)` on it
1. `/datum/component/var/list/signal_procs` (private)
    * Associated lazy list of signals -> callbacks that will be run when the parent datum recieves that signal
1. `/datum/component/var/datum/parent` (protected, read-only)
    * The datum this component belongs to

### Procs

1. `/datum/proc/GetComponent(component_type(type)) -> datum/component?` (public, final)
    * Returns a reference to a component of component_type if it exists in the datum, null otherwise
1. `/datum/proc/GetComponents(component_type(type)) -> list` (public, final)
    * Returns a list of references to all components of component_type that exist in the datum
1. `/datum/proc/GetExactComponent(component_type(type)) -> datum/component?` (public, final)
    * Returns a reference to a component whose type MATCHES component_type if that component exists in the datum, null otherwise
1. `GET_COMPONENT(varname, component_type)` OR `GET_COMPONENT_FROM(varname, component_type, src)`
    * Shorthand for `var/component_type/varname = src.GetComponent(component_type)`
1. `/datum/proc/AddComponent(component_type(type), ...) -> datum/component`  (public, final)
    * Creates an instance of `component_type` in the datum and passes `...` to it's `New()` call
    * Sends the `COMSIG_COMPONENT_ADDED` signal to the datum
    * All components a datum owns are deleted with the datum
    * Returns the component that was created. Or the old component in a dupe situation where `COMPONENT_DUPE_UNIQUE` was set
1. `/datum/proc/ComponentActivated(datum/component/C)` (abstract)
    * Called on a component's `parent` after a signal recieved causes it to activate. `src` is the parameter
    * Will only be called if a component's callback returns `TRUE`
1. `/datum/proc/TakeComponent(datum/component/C)` (public, final)
    * Properly transfers ownership of a component from one datum to another
    * Singals `COMSIG_COMPONENT_REMOVING` on the parent
    * Called on the datum you want to own the component with another datum's component
1. `/datum/proc/SendSignal(signal, ...)` (public, final)
    * Call to send a signal to the components of the target datum
    * Extra arguments are to be specified in the signal definition
1. `/datum/component/New(datum/parent, ...)` (protected, virtual)
    * Forwarded the arguments from `AddComponent()`
1. `/datum/component/Destroy()` (virtual)
    * Sends the `COMSIG_COMPONENT_REMOVING` signal to the parent datum if the `parent` isn't being qdeleted
    * Properly removes the component from `parent` and cleans up references
1. `/datum/component/proc/InheritComponent(datum/component/C, i_am_original(boolean))` (abstract)
    * Called on a component when a component of the same type was added to the same parent
    * See `/datum/component/var/dupe_mode`
    * `C`'s type will always be the same of the called component
1. `/datum/component/proc/OnTransfer(datum/new_parent)` (abstract)
    * Called before the new `parent` is assigned in `TakeComponent()`, after the remove signal, before the added signal
    * Allows the component to react to ownership transfers
1. `/datum/component/proc/_RemoveNoSignal()` (private, final)
    * Internal, clears the parent var and removes the component from the parents component list
1. `/datum/component/proc/RegisterSignal(signal(string), proc_ref(type), override(boolean))` (protected, final) (Consider removing for performance gainz)
    * Makes a component listen for the specified `signal` on it's `parent` datum.
    * When that signal is recieved `proc_ref` will be called on the component, along with associated arguments
    * Example proc ref: `.proc/OnEvent`
    * If a previous registration is overwritten by the call, a runtime occurs. Setting `override` to TRUE prevents this
    * These callbacks run asyncronously
    * Returning `TRUE` from these callbacks will trigger a `TRUE` return from the `SendSignal()` that initiated it
1. `/datum/component/proc/ReceiveSignal(signal, ...)` (virtual)
    * Called when a component recieves any signal and is enabled
    * Default implementation looks if the signal is registered and runs the appropriate proc

## Examples
    Material Containers: #29268 (Too many GetComponent calls, but not bad)
    Slips: #29339
    Powercells: (TODO)

### CODE REVIEW

**Comment by MrStonedOne** [comment]
I'll open up a new pr once this is merged, this doesn't resolve the issue.

File: code/datums/components/README.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+## Concept
+
+Loosely adapted from /vg/. This is an entity component system for adding behaviours to datums when inheritance doesn't quite cut it. By using signals and events instead of direct inheritance, you can inject behaviours without hacky overloads. It requires a different method of thinking, but is not hard to use correctly. If a behaviour can have application across more than one thing. Make it generic, make it a component. Atom/mob/obj event? Give it a signal, and forward it's arguments with a `SendSignal()` call. Now every component that want's to can also know about this happening.
+
+### In the code
+
+#### Slippery things
+
+At the time of this writing, every object that is slippery overrides atom/Crossed does some checks, then slips the mob. Instead of all those Crossed overrides they could add a slippery component to all these objects. And have the checks in one proc that is run by the Crossed event
```
Comment (Jalleo) [inline code/datums/components/README.md]: This is better to talk and collaborate about as a project and work on it rather than leaving to dust away here.

**Comment by outbools** [comment]
Actually, HOLY CRAB
That's obvious we need to document EVERY COMPONENT since the majority of them meant to be reused! It should be forced on contributors like changelog, maybe automated same way.

Example:
It would be fucking disgusting if someone ever create one component instead of two existing which do the same thing combined and start fucking slippery slope with that.

**Comment by Iamgoofball** [comment]
MSO if the issue is documentation this resolves it pretty fucking good

File: code/datums/components/README.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    * Singals `COMSIG_COMPONENT_REMOVING` on the parent
+    * Called on the datum you want to own the component with another datum's component
+1. `/datum/proc/SendSignal(signal, ...)` (public, final)
+    * Call to send a signal to the components of the target datum
+    * Extra arguments are to be specified in the signal definition
+1. `/datum/component/New(datum/parent, ...)` (protected, virtual)
+    * Forwarded the arguments from `AddComponent()`
+1. `/datum/component/Destroy()` (private)
+    * Use `RemoveComponent()` instead of `qdel()` to properly delete a component
+    * `qdel()` still works but it suppresses the `COMSIG_COMPONENT_REMOVING` signal and is not intended behavior
```
Comment (MrPerson) [inline code/datums/components/README.md]: This is frankly incorrect. You should fix this behavior to be more standard with everything else ie qdel() to delete it.

Comment (Cyberboss) [inline code/datums/components/README.md]: Yeah, I'm trying to figure out how without adding another var, but it seems unlikely

File: code/datums/components/README.md (hunk 1 above)
Comment (Supermichael777) [inline code/datums/components/README.md]: implement this to show how its done then list it step by step as a practical example

Comment (Cyberboss) [inline code/datums/components/README.md]: #29339

**Comment by MrPerson** [comment]
What do you mean you can't figure out how? Just move all the logic from RemoveComponent to Destroy.  It's not complicated.

**Comment by MrStonedOne** [comment]
Its not documentation I want goof, it's that the original PR did not describe what the pr did in the pr body.

Pretending this is a solution is part of the problem.

I want cyberboss to:

1. Edit a proper description into that pr, 

1. Add a proper description to this pr.

1. Acknowledge that he messed up by not doing so.

1. Apologize for not doing

1. State he will make a concerted effort to give better descriptions in the future.

**Comment by Iamgoofball** [comment]
do you have a time machine? because the past is the past

**Comment by MrStonedOne** [comment]
This isn't about the past, this is about the future, and the fact that @cyberboss seems to think slapping some readme document in the repo is a adequate replacement for the basic requirement that you ***DETAIL WHAT THE FUCK YOUR PR IS*** tells me nothing has been learned, and this issue will come up in the future if nothing is done to change it.

**Comment by MrStonedOne** [comment]
like @Iamgoofball, this pr itself doesn't even have a fucking description.

Like that makes my point clearer than anything else could.

**Comment by Cyberboss** [comment]
Updated, happy now?

**Comment by Cyberboss** [comment]
Ugh too much drama nevermind

**Comment by outbools** [comment]
@MrStonedOne
>Acknowledge that he messed up by not doing so.
>Apologize for not doing so
>State he will make a concerted effort to give better descriptions in the future

Wow this amazing managment over VOLONTEER team, isn't it?

**Comment by Quantomicus** [comment]
Can your repo get any more of a joke? Please.

**Comment by TeknoKot** [comment]
ONE NAME ABOVE ALL, ALL OTHERS WILL FALL

**Comment by PJB3005** [comment]
Honestly this just seems like mso doubling down instead of being reasonable again. Happens a lot.

TG you genuinely need to get yourself together. I'm not kidding when I say your codebase is a joke. This is just sad.

**Comment by Cyberboss** [comment]
@PJB3005 convinced me

**Comment by Qustinnus** [comment]
god bless the queen

**Comment by PJB3005** [comment]
👏

**Comment by outbools** [comment]
papa bless

**Comment by MrStonedOne** [comment]
>I want cyberboss to:

- [ ] Edit a proper description into that pr,

- [x] Add a proper description to this pr.

- [ ] Acknowledge that he messed up by not doing so.

- [ ] Apologize for not doing so

- [ ] State he will make a concerted effort to give better descriptions in the future.

1 out of 5 so far. If I don't see all 5 met when this PR is merged, I will open a revert of this pr and the other, and then quick merge it using @tgstation-server.

I'm not fucking playing around here @Cyberboss.

**Comment by outbools** [comment]
@MrStoneOne so if he's not going to apologize we cant work on DCS?

**Comment by clusterfack** [comment]
What sort of thing do you believe he should apologize for?

**Comment by MrStonedOne** [comment]
Repeatedly making description-less prs against the policy.

Trying to scapegoat by blaming it on git, like having a readme would have made a difference

Trying to scapegoat it by opening up a rage quit revert PR and calling the basic requirement that you detail what your PR does "Drama".

I'm not asking for an essay, but I haven't seen so much as a "whoops" or a "my bad".

Instead I get:

>Updated, happy now?

**Comment by clusterfack** [comment]
Unless he is merging his own PRs that violate your contributor guidelines, the ones at fault are those maintainers who are merging his PRs without enforcing your contributor guidelines.

**Comment by MrStonedOne** [comment]
I put as much blame on the coder for making a bad PR as I put on the maintainer for merging a bad PR.

**Comment by clusterfack** [comment]
Are you? I must have missed seeing you demand a public apology from the maintainer who merged it nor a promise to no longer merge things that violate contributor guidelines.

**Comment by MrStonedOne** [comment]
Except oranges hasn't attempted to scapegoat this issue.

This has only gotten as far as it has because cyberboss has refused to admit he fucked up.

Read here: https://github.com/tgstation/tgstation/pull/29317

**Comment by outbools** [comment]
You just harassing a guy who made you 1k refactor, bugfix and feature PRs for free, sacrificing his own time just to let you stomp on his work for thing called "guideline"

NO ONE OWE YOU APOLOGIES

THIS IS FUCKED UP BECAUSE YOU OPENING REVERT PRs AND ASKING FOR APOLOGIES BECAUSE OF IT.

You messing up several people work and actually stops us from contributing to DCS because we dont fucking know how it would end up. Please, cease this drama and let people work.

**Comment by ChangelingRain** [comment]
Volunteer contributors are not required to give apologies for failing to follow the guidelines.
Yes, they have to. No, they don't need to apologize, they just need to fix their mistake.

**Comment by Qustinnus** [comment]
You realize that doing this doesn't just punish Cyberboss for whatever thing you're so angry about right?  A lot of us want to work with this system and you going "haha ill revert it anyway" just blocks all of us from working on the systems we want to work on.

Honestly I get that you're mad at Cyber but seeing  what he has done in the past for TG I think you both just need to chill, I think Cyber got the message by now and if his descriptions aren't in line in the future then you can find some other way to punish him for this. But honestly this just wont work, and I dont think he should have to apologize for his previous PRs.

**Comment by outbools** [comment]
>they don't need to apologize
@ChangelingRain and now look at MSO checklist, dear maintainer. Point 4.

This is why situation fucked up:

- Cyberboss opened a PR without complete description

- He's and I described to reviewing maintainer what this PR does.

- PR got merged

- MSO opened a revert stating:
>PR didn't explain what the fuck this is.
>So I hate it.

- PR was already merged so Cyberboss provided description (good, excellent one) with this PR. This PR was made BECAUSE of MSO request.

- MSO never stated clear that Cyberboss should update previous PR description until yesterday

- He also asking cyberboss to APOLOGIZE BECAUSE HE CANT UNDERSTAND HIS VAGUE REQUESTS. He's just keep being edgy and rude.

So who's should be blamed here??

**Comment by outbools** [comment]
@MrStonedOne
This is fucking stupid - can you just NOT bring your relations and miscommunication with YOUR MAINTAINERS to public? Don't you have a fucking private discord chat for staff to talk with each other. Im fucking SURPRISED with level of miscommunication you got on /tg/ and you still pushing this kind of behaviour through your codebase.

**Comment by MrStonedOne** [comment]
@outbools we do not have a private discord, we do not have a private forum page, and if there ***is*** a private irc room, I'm not allowed in it.

**Comment by outbools** [comment]
Wow that's amazing so you can't even communicate with each other without arguing on public. Amazing!

**Comment by Iamgoofball** [comment]
@hornygranny 

please help

**Comment by MrStonedOne** [comment]
>He's and I described to reviewing maintainer what this PR does.

You and him mentioned some vague shit relating to the general concept of ECS/components, and something about subsystems, but neither of you detailed what that PR in in of itself does, how it implemented the broad ECS concept.

>PR was already merged so Cyberboss provided description (good, excellent one) with this PR. This PR was made BECAUSE of MSO request.

>MSO never stated clear that Cyberboss should update previous PR description until yesterday

The description wasn't added to this pr until AFTER I added that list.

This PR is actually still breaking this the rules. it does not state what this pr does, it states what the other pr does, as the body of this pr.

***This pr actually makes quite a few changes that are not documented or explained in the PR body.***

**Comment by MrStonedOne** [comment]
I have now unchecked point number 2 on my request list.

we are now at 0 out of 5.

**Comment by outbools** [comment]
local host drunk with power and goes insane

10/10 would unsubscribe from updates again

**Comment by outbools** [comment]
Please, probably an autistic person (because ignoring to speak directly and actually discuss things with Cyberboss first) , can you have a normal tet-a-tet conversation with Cyberboss to see why you boss misunderstanding each other so much?

**Comment by d3athrow** [comment]
>porting vg code

How dare you

**Comment by PJB3005** [comment]
Wait @MrStonedOne but I never apologized for naming the lighting PR "rebase
to /vg/"

**Comment by outbools** [comment]
i want PJB to: 

- [ ] Edit a proper description into that pr,
- [ ] Add a proper description to this pr.
- [ ] Acknowledge that he messed up by not doing so.
- [ ] Apologize for not doing so
- [ ] State he will make a concerted effort to give better descriptions in the future

image

**Comment by dannno** [comment]
haha whatta fuck is this, man

can you cool it you fucking idiots

**Comment by Wyzack** [comment]
Is making him lick your boots and apologize really necessary?

**Comment by dannno** [comment]
It will never be necessary and it's a complete embarrassment that he's trying to demand it to begin with

**Comment by Cyberboss** [comment]
The previous PR's description has been updated and that's all I'm doing

**Comment by MrStonedOne** [comment]
You still need to document this pr properly.

Something along the lines of `Adds the documentation that was missing as well as something something bug fix something something qdel`.

And I still want to see at least point 3 addressed. You've been scapegoating this entire thing, so I want to at least see you admit or acknowledge fault. even if it's just a simple `my bad`.

**Comment by ChangelingRain** [comment]
I'm getting kind of tempted to just merge this and accelerate the drama, since it's a perfectly functional pr and system being helped back by someone wanting the author to acknowledge fault.

**Comment by MrStonedOne** [comment]
This PR still has no valid description describing what it does. The description provided is for another PR, and this one has changes that aren't documented

**Comment by optimumtact** [comment]
lol what is going on here

**Comment by Qustinnus** [comment]
only god knows

**Comment by optimumtact** [comment]
we have a private irc channel?

**Comment by optimumtact** [comment]
I ticked all 5 of MSO's boxes so I can merge this now

**Pull request merged**